plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -344,11 +344,13 @@ maybeAddTySubst tn rhs = do
 -- | Is this a an utterly trivial term which might as well be inlined?
 trivialTerm :: Term tyname name uni fun a -> Bool
 trivialTerm = \case
-    Builtin{}  -> True
-    Var{}      -> True
+    Builtin{}      -> True
+    Var{}          -> True
     -- TODO: Should this depend on the size of the constant?
-    Constant{} -> True
-    _          -> False
+    Constant{}     -> True
+    TyAbs _ _ _ t  -> trivialTerm t
```

Comment:
This needs a note. The decision making here is actually quite complicated. We need to think about two things: saving work, and saving space.

In terms of saving work, I think this is fine. We will create the outermost lambda value at each use site instead of once, but that's a pretty small amount of work that I'm fairly okay with rounding to 0 (but we should document that in a note!). Indeed, several of the other cases also require us to do value creation at each use site, although creating lambda values is probably the most expensive kind (not by much).

In terms of saving space, this is riskier. We will duplicate the whole term at each use site. The other cases all relate to terms which are approximately the same size as a variable (except the constant case! which has a worrying TODO on it...). We could make this much safer by actually doing something more like what the GHC inliner does, and conditionally inlining things at call sites if they're fully applied. But this would be much more complicated...

I think for starters this could do with:
- A note
- Splitting up this function into two that represent the two decisions that we're making. Something like `costIsAcceptable` and `sizeIsAcceptable`. That way it's clearer what they're checking for and we could in principle use them separately if that's appropriate elsewhere. Then `postInlineUnconditional` can require that both hold.

Suggested implementations:
```
costIsAcceptable = \case
  Builtin{} -> True
  Var{} -> True
  Constant{} -> True
  LamAbs{} -> True
  TyAbs{} -> True
  Error{} -> True
  IWrap _ _ _ t -> costIsAcceptable t
  Unwrap _ t -> costIsAcceptable t

  Apply{} -> False
  TyInst{} -> False
  Let{} -> False

sizeIsAcceptable = \case
  Builtin{} -> True
  Var{} -> True
  -- These are the two dubious cases, note here
  LamAbs _ _ _ t -> sizeIsAcceptable t
  TyAbs _ _ _ t -> sizeIsAcceptable t
  Error{} -> True
  
  Constant{} -> False --- I actually think inlining constants is pretty dubious 
  -- Arguably we could allow these two...
  IWrap _ _ _ t -> False
  Unwrap _ t -> False
  Apply{} -> False
  TyInst{} -> False
  Let{} -> False
```